massa-execution-worker/src/execution.rs

```diff
@@ -1221,6 +1225,122 @@ impl ExecutionState {
         }
     }
 
+    fn execute_deferred_call(
+        &self,
+        call: DeferredCall,
+    ) -> Result<DeferredCallExecutionResult, ExecutionError> {
+        let mut result = DeferredCallExecutionResult::new(&call);
+
+        let snapshot;
+        // prepare the current slot context for executing the operation
+        let bytecode;
+        {
+            let mut context = context_guard!(self);
+            snapshot = context.get_snapshot();
+            // acquire write access to the context
+            // let mut context = context_guard!(self);
+
+            // Set the call stack
+            // This needs to be defined before anything can fail, so that the emitted event contains the right stack
+            context.stack = vec![
+                ExecutionStackElement {
+                    address: call.sender_address,
+                    coins: call.coins,
+                    owned_addresses: vec![call.sender_address],
+                    operation_datastore: None,
+                },
+                ExecutionStackElement {
+                    address: call.target_address,
+                    coins: call.coins,
+                    owned_addresses: vec![call.target_address],
+                    operation_datastore: None,
+                },
+            ];
+
+            // Ensure that the target address is an SC address
+            // Ensure that the target address exists
+            if let Err(err) = context.check_target_sc_address(call.target_address) {
+                context.reset_to_snapshot(snapshot, err.clone());
+                context.deferred_call_fail_exec(&call);
+                return Err(err);
+            }
+
+            // credit coins to the target address
+            if let Err(err) =
+                context.transfer_coins(None, Some(call.target_address), call.coins, false)
+            {
+                // coin crediting failed: reset context to snapshot and reimburse sender
+                let err = ExecutionError::RuntimeError(format!(
+                    "could not credit coins to target of deferred call execution: {}",
+                    err
+                ));
+
+                context.reset_to_snapshot(snapshot, err.clone());
+                context.deferred_call_fail_exec(&call);
+                return Err(err);
+            }
+
+            // quit if there is no function to be called
+            if call.target_function.is_empty() {
+                return Err(ExecutionError::RuntimeError(
```

Comment:
this error does not trigger deferred_call_fail and snapshot reset.

Reply:
now properly wrapped in a closure to trigger fails function